otherlibs/unix/realpath.c

```diff
@@ -0,0 +1,43 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                         The OCaml programmers                          */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/alloc.h>
+#include <caml/fail.h>
+#include "unixsupport.h"
+
+#ifdef HAS_REALPATH
+
+CAMLprim value unix_realpath (value p)
+{
+  CAMLparam1 (p);
+  char *r;
+  value rp;
+
+  caml_unix_check_path (p, "realpath");
+  r = realpath (String_val (p), NULL);
```

Comment:
I don't know if there are any supported platforms that don't support this, but since this relies on the POSIX 2008 behavior of realpath allocating its own buffer when passed NULL, should configure check that the found realpath is new enough (or the GNU one)?

Reply:
I don't know either and have no idea how to test that via `configure`.

Reply:
So, it seems that in [POSIX 2001](https://pubs.opengroup.org/onlinepubs/009695399/functions/realpath.html) a NULL pointer for `resolved_name` is implementation defined. At least for `glibc` that meant (prior to 2.3) returning `EINVAL` (cf. the [linux manpage](https://linux.die.net/man/3/realpath)).

Besides the latest [`netbsd`](https://man.netbsd.org/NetBSD-9.1-STABLE/i386/realpath.3), [`openbsd`](https://man.openbsd.org/OpenBSD-6.8/man3/realpath.3) and [`freebsd`](https://www.freebsd.org/cgi/man.cgi?query=realpath&apropos=0&sektion=3&manpath=FreeBSD+12.2-stable&arch=default&format=html) all seem to support the [POSIX 2008](https://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html) version.

Given the above. I propose to keep this as such.